### ISSUE TITLE
[OpenMP] AMDGPU Memory access fault

### ISSUE THREAD

**Issue by ye-luo** [llvm/llvm-project#63221]
The reproducer can reliably fail on gfx906 but not gfx90a.
Using miniqmc https://github.com/ye-luo/miniqmc
```
mkdir build_r7_clangNightly_offload_MP; cd build_r7_clangNightly_offload_MP
cmake -DCMAKE_CXX_COMPILER=clang++ -DENABLE_OFFLOAD=ON -DQMC_MIXED_PRECISION=ON -DQMC_GPU_ARCHS=gfx906 ..
make -j32 test_distance_table
ctest -VV -R test_distance_table
```
failure error
```
AMDGPU fatal error 1: Memory access fault by GPU 1 (agent 0x563f70b34d40) at virtual address (nil). Reasons: Page not present or supervisor privilege, Write access to a read-only page
```
offending offload region https://github.com/ye-luo/miniqmc/blob/25aa5b4bd47032c6738441d4296f609e08f5f91a/src/Particle/SoaDistanceTableAAOMPTarget.h#L337

**Comment by jdoerfert** [comment]
I traced this down to a backend bug. https://reviews.llvm.org/D145329 impacts it but doesn't fix it. In fact, before https://reviews.llvm.org/D145329 I can produce correct results if I stop early enough in the middle or backend, with it I mostly crash for anything that is not O0 (in the backend).

**Comment by arsenm** [comment]
Not failing for me on gfx1031 on main

**Comment by ye-luo** [comment]
```
commit f914208c4388780e24ed3a1ab2e170c53129b2a1
Author: Johannes Doerfert <johannes@jdoerfert.de>
Date:   Tue Jul 18 16:33:31 2023 -0700

    [OpenMP][NFCI] Avoid storing non-constant values in ICV
    
    If we store a constant in an ICV it is easier for the optimizer to
    propagate it. Since we often use the full block for the thread limit and
    the parallel team size, we can instead replace that dynamic value with a
    constant that otherwise cannot occur, here 0.
```
made the test passing again. @jdoerfert is it because this fix bypasses the backend bug?

**Comment by jdoerfert** [comment]
> ```
> commit f914208c4388780e24ed3a1ab2e170c53129b2a1
> Author: Johannes Doerfert <johannes@jdoerfert.de>
> Date:   Tue Jul 18 16:33:31 2023 -0700
> 
>     [OpenMP][NFCI] Avoid storing non-constant values in ICV
>     
>     If we store a constant in an ICV it is easier for the optimizer to
>     propagate it. Since we often use the full block for the thread limit and
>     the parallel team size, we can instead replace that dynamic value with a
>     constant that otherwise cannot occur, here 0.
> ```
> 
> made the test passing again. @jdoerfert is it because this fix bypasses the backend bug?

Yes. We optimize better with that commit and generate better (=different) code that does not trigger the bug.
